### ISSUE TITLE
Unable to process all vocabs

### ISSUE THREAD

**Issue by jojipy** [elKuston/WaniKaniAnkiAddon#1]
Hello,
I've been looking for a solution to recall my WaniKani vocabulary to Anki, and stumbled upon your plugin, which seemed to work until a couple of months ago. I was waiting for the problem to resolve itself, but it didn't, which is why I'm submitting this issue.

After reinstalling my operating system, I had to re-sync all my WaniKani vocabulary with a backup of my original Anki decks, after which your plugin suddenly stopped working. After waiting for the request limit to reset, it returns the error "Sleep length must be non-negative". I managed to overcome this issue by manually setting the sleep length in the source-code to 30 seconds. However after loading around 3000 vocabulary another issue pops up returning "list index out of range".

If you need any other further information, I'd be happy to provide it.

**Comment by elKuston** [comment]
Hello! I'm sorry to hear about your inconvenience. There is a newer version i've been testing which may (or may not) correct your issue, tho I haven't released it yet because there is an error that appears sporadically which I couldn't manage to resolve (and, unfortunately, I don't have too much time right now). It is an error that is displayed to the user, but has no actual effect as far as I have noticed, so you should not worry about it. I have tested it with about 1600 vocabulary and seems to be working fine, so I hope it will also work for you. If it is okay, I will attach the __init__.py file (zipped, as GitHub doesn't allow .py attached files) in this message and will ask you to replace your current script with the new one (given you changed the source code I imagine you know how to do this, but you can ask for help in case you don't) and run Anki from the console, so that if an error occurs you can share with me the full error. To do this you just have to open your anki installation folder and run anki-console.bat. If the script fixes your error, I will later release it as an update.
Hope this helps.
Thank you.
[__init__.zip](https://github.com/elKuston/WaniKaniAnkiAddon/files/11451231/__init__.zip)

Edit: Forgot to mention, this new script uses a new card type that includes the "parts of the speech" section, which I thought could be useful. It should not duplicate previous cards in the collection. Also, daily load times should be shorter, as this version loads only vocabulary from lessons you've taken since the last sync, rather than loading all vocab modified since then (which includes reviewes) as the old one did. Some recently-learnt vocab may not be loaded, depending on your time zone, in which case it will be loaded in a future start (That's something i'll be working on for future releases).

**Comment by jojipy** [comment]
Hello, thank you for your fast reply. The issue with the "Sleep length must be non-negative" still persists. As before I manually replaced it with a set time of 30 seconds, after which it worked normally, until I encountered the same issue again ("list index out of range"). I'll attach the log containing both errors (with my temporary fix of setting the line with the sleep time to 30 seconds; other than that line no other changes were made). I hope it helps.
[log.txt](https://github.com/elKuston/WaniKaniAnkiAddon/files/11455091/log.txt)

A quick side note: I looked through the log, and I found something regarding the sleep time in line 207 of the log. The other error is right at the end.

**Comment by elKuston** [comment]
Okay, I have found the problem regarding the "index out of range" error: I assumed that every vocab had, at least, one pronunciation audio, but apparently in levels higher than mine some do not have it (such as  [絞める](https://www.wanikani.com/vocabulary/%E7%B5%9E%E3%82%81%E3%82%8B) ). I will correct this and the sleep time error and upload a new version over the weekend. Thanks for you patience and your help.

**Comment by elKuston** [comment]
Hi @jojipy. I think I have solved the issues, but given I don't have access with my API key to the "conflictive" vocab, I would like you to test the attached script and verify whether the error is solved before actually uploading it, if that's ok for you.
Thanks.
[__init__.zip](https://github.com/elKuston/WaniKaniAnkiAddon/files/11471402/__init__.zip)

**Comment by jojipy** [comment]
Hi @elKuston,
The loading of the vocabulary works without any problems, however after having loaded them all in, I get a new error:
"
[...]
loading subject 5291; 閣僚
loading subject 5401; 官僚的
Traceback (most recent call last):
  File "concurrent.futures.thread", line 58, in run
  File "aqt.operations", line 239, in wrapped_op
  File "/home/user/.local/share/Anki2/addons21/710584815/__init__.py", line 180, in import_vocab_from_wanikani
    add_cards_anki(col, all_subjects_mapped)
  File "/home/user/.local/share/Anki2/addons21/710584815/__init__.py", line 152, in add_cards_anki
    anki_note = subject_to_anki_note(col, deck['id'], model, subject)
  File "/home/user/.local/share/Anki2/addons21/710584815/__init__.py", line 112, in subject_to_anki_note
    download(col, audio_filename, subject['audio'])
  File "/home/user/.local/share/Anki2/addons21/710584815/__init__.py", line 90, in download
    resp = client.get(url)
  File "anki.httpclient", line 64, in get
  File "requests.sessions", line 600, in get
  File "requests.sessions", line 573, in request
  File "requests.sessions", line 484, in prepare_request
  File "requests.models", line 368, in prepare
  File "requests.models", line 439, in prepare_url
requests.exceptions.MissingSchema: Invalid URL 'None': No scheme supplied. Perhaps you meant http://None?
"

Edit: I just noticed that a deck has been generated, however it seems incomplete (missing around 500 vocabulary). After restarting Anki it attempts to reload all vocabulary (including the ones it has successfully added to the deck). I presume this has something to do with the error I mentioned

**Comment by elKuston** [comment]
I have added some extra null checks that should resolve the issue. I have tested it with non-audio vocabs and it worked on my device, so I hope it also works on yours. Regarding the incomplete deck, it is an expected behavious to redownload all the cards, because after a crash it does not keep track of which cards were actually added (it should not duplicate any card however). I plan to add fault tollerance in future releases. 
I hope the attached script solves the issue this time.
Thanks for your patience and your help.

[__init__.zip](https://github.com/elKuston/WaniKaniAnkiAddon/files/11472142/__init__.zip)
